File: engine/Kogase.Engine/Models/Device.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Kogase.Engine.Models
{
    public class Device
    {
        [Key]
        public int Id { get; set; }
        
        [Required]
        public int ProjectId { get; set; }
        
        [ForeignKey("ProjectId")]
        public Project? Project { get; set; }
        
        [Required]
        [StringLength(100)]
        public string DeviceId { get; set; } = null!;
        
        [Required]
        [StringLength(50)]
        public string Platform { get; set; } = null!;
        
        [Required]
        [StringLength(50)]
        public string OsVersion { get; set; } = null!;
        
        [Required]
        [StringLength(50)]
        public string AppVersion { get; set; } = null!;
        
        [Required]
        public DateTime FirstSeen { get; set; }
        
        [Required]
        public DateTime LastSeen { get; set; }
        
        public string? Country { get; set; }

        public ICollection<Session> Sessions { get; set; } = new List<Session>();
        public ICollection<Event> Events { get; set; } = new List<Event>();
    }
} 

File: backend/models/device.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Device represents a unique player device
type Device struct {
	ID         uuid.UUID      `json:"id" gorm:"type:uuid;primary_key"`
	ProjectID  uuid.UUID      `json:"project_id" gorm:"type:uuid;not null"`
	DeviceID   string         `json:"device_id" gorm:"not null"`            // Client-generated device identifier
	Platform   string         `json:"platform" gorm:"not null"`             // iOS, Android, Windows, etc.
	OSVersion  string         `json:"os_version" gorm:"not null"`           // e.g., "10.0", "Android 11"
	AppVersion string         `json:"app_version" gorm:"not null"`          // Game version
	FirstSeen  time.Time      `json:"first_seen" gorm:"not null"`           // First session timestamp
	LastSeen   time.Time      `json:"last_seen" gorm:"not null"`            // Last session timestamp
	IPAddress  string         `json:"ip_address,omitempty" gorm:"not null"` // Hashed/anonymized IP address
	Country    string         `json:"country,omitempty"`                    // Country based on IP (optional)
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
	DeletedAt  gorm.DeletedAt `json:"-" gorm:"index"`
	Project    Project        `json:"-" gorm:"foreignKey:ProjectID"`
	Events     []Event        `json:"events,omitempty" gorm:"foreignKey:DeviceID"`
}

// BeforeCreate will set a UUID rather than numeric ID
func (device *Device) BeforeCreate(_ *gorm.DB) error {
	if device.ID == uuid.Nil {
		device.ID = uuid.New()
	}

	// Set first and last seen if not set
	now := time.Now()
	if device.FirstSeen.IsZero() {
		device.FirstSeen = now
	}
	if device.LastSeen.IsZero() {
		device.LastSeen = now
	}

	return nil
}
